pkg/sql/colexec/mergejoiner.go

```diff
@@ -249,30 +249,126 @@ func NewMergeJoinOp(
+	// width mismatch, we need to cast one of the vectors to another and use
+	// the cast vector for equality check.
+	// Note that there are type pairs of other numeric types that are valid for
+	// equality check (e.g. FLOAT vs INT4, or INT8 vs DECIMAL, etc), but we
+	// don't plan merge joins in such scenarios - instead, we have a cross join
+	// with a filter on top, and the cross joins are handled by falling back to
+	// rowexec.hashJoiner since we don't have a cross join operator at the
+	// moment (#46205).
+
+	// Make a copy of types and orderings to be sure that we don't modify
+	// anything unwillingly.
+	actualLeftTypes, actualRightTypes := append([]*types.T{}, leftTypes...), append([]*types.T{}, rightTypes...)
+	actualLeftOrdering := make([]execinfrapb.Ordering_Column, len(leftOrdering))
+	actualRightOrdering := make([]execinfrapb.Ordering_Column, len(rightOrdering))
+	copy(actualLeftOrdering, leftOrdering)
+	copy(actualRightOrdering, rightOrdering)
+	// Iterate over each equality column and check whether a cast is needed. If
+	// it is needed for some column, then a cast operator is planned on top of
+	// the input from the corresponding side and the types and ordering are
+	// adjusted accordingly. We will also need to project out that temporary
+	// column, so a simple project will be planned below.
+	needProjection := false
+	for i := range leftOrdering {
+		leftColIdx := leftOrdering[i].ColIdx
+		rightColIdx := rightOrdering[i].ColIdx
+		leftType := leftTypes[leftColIdx]
+		rightType := rightTypes[rightColIdx]
+		if leftType.Identical(rightType) {
+			// The types are the same, so no cast is needed.
+			continue
+		}
+		var err error
+		if leftType.Family() != types.IntFamily || rightType.Family() != types.IntFamily {
+			return nil, errors.AssertionFailedf(
+				"attempting to create a merge joiner on equality columns of different non-integer types",
+			)
+		}
+		// There is a hierarchy of valid casts:
+		//   INT2 -> INT4 -> INT8
+		// and the cast is valid if 'fromType' is mentioned before 'toType'
+		// in this chain.
+		castLeftToRight := false
+		switch leftType.Width() {
+		case 16:
+			castLeftToRight = true
+		case 32:
+			castLeftToRight = !rightType.Identical(types.Int2)
+		}
+		if castLeftToRight {
+			castColumnIdx := len(actualLeftTypes)
+			left, err = GetCastOperator(unlimitedAllocator, left, int(leftColIdx), castColumnIdx, leftType, rightType)
+			if err != nil {
+				return nil, err
+			}
+			actualLeftTypes = append(actualLeftTypes, rightType)
+			actualLeftOrdering[i].ColIdx = uint32(castColumnIdx)
+		} else {
+			castColumnIdx := len(actualRightTypes)
+			right, err = GetCastOperator(unlimitedAllocator, right, int(rightColIdx), castColumnIdx, rightType, leftType)
+			if err != nil {
+				return nil, err
+			}
+			actualRightTypes = append(actualRightTypes, leftType)
+			actualRightOrdering[i].ColIdx = uint32(castColumnIdx)
+		}
+		needProjection = true
+	}
```

Comment:
I'm surprised to see this logic in the merge joiner and not the flow creator, but maybe it's not a big deal. Philosophically, I think we're split between wanting operator constructors to create a fully fledged flow of operators that matches a spec, and wanting constructors to be as simple as possible and putting the complexity in the flow creator, right?